server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -468,12 +475,15 @@ public void updateShardState(final ShardRouting newRouting,
                     if (resyncStarted == false) {
                         throw new IllegalStateException("cannot start resync while it's already in progress");
                     }
-                    indexShardOperationPermits.asyncBlockOperations(
-                        30,
-                        TimeUnit.MINUTES,
+                    bumpPrimaryTerm(newPrimaryTerm,
                         () -> {
                             shardStateUpdated.await();
+                            assert pendingPrimaryTerm == newPrimaryTerm :
```

Comment:
1、in method updateShardState ,because newPrimaryTerm != pendingPrimaryTerm so the code run into bumpPrimaryTerm method 
![image](https://user-images.githubusercontent.com/44155914/65228772-a917c880-dafd-11e9-80d4-d1f44c7f6c33.png)

2、in bumpPrimaryTerm，it will set pendingPrimaryTerm = newPrimaryTerm
![image](https://user-images.githubusercontent.com/44155914/65228895-e419fc00-dafd-11e9-9177-620315ec849e.png)

3、but  before bumpPrimaryTerm set pendingPrimaryTerm = newPrimaryTerm ，onBlocked will run first，in which will assert pendingPrimaryTerm == newPrimaryTerm
![image](https://user-images.githubusercontent.com/44155914/65229035-22afb680-dafe-11e9-9f3e-91749a87da57.png)
![image](https://user-images.githubusercontent.com/44155914/65229109-470b9300-dafe-11e9-9683-7d8a607fd774.png)